File: BinarySearch/BinarySearch.cpp
#include <stdio.h>
#include <iostream>

int binarySearch(int arr[], int len, int target) {

    // Initialize the variables to start the search
    int start = 0; int end = len - 1;

    // Loop until the value is found or the array is exhausted
    while (start <= end)
    {
        
        // find the midpoint of the array 
        int mid = (start + end) / 2;

        // Target is found 
        if (arr[mid] == target) {
            return mid;
        } else if (arr[mid] < target)  // The target is on the right half of the list
        {
            start = mid + 1;
        } else {                       // The target is on the left half of the list
            end = mid - 1;
        }
        
    }

    // Target was not found
    return -1;

}

// Demonstrates the BinarySearch algorithm in C++
int main() {
    // Create the start array 
    int arr[100];
    int target = 60;  // Search Target
    int idx;          // Idx of target

    // Load up the array with starting values 
    for(int i = 0; i < 100; i++) {
        arr[i] = i + 1;
    }

    // Print the array out
    for(int i = 0; i < 100; i++) {
        std::cout << "Index: " << i << ", Value: " << arr[i] << "\n";
    }

    std::cout << "\nSearching for target: " << target;
    idx = binarySearch(arr, 100, target);
    
    if (idx != -1) {
        std::cout << "\nFound " << target << " at position " << idx << "\n";
    } else {
        std::cout << "\nTarget was not found in the list\n";
    }

    return 0;
}

File: BinarySearch/binarySearch.py

def main():
    
    # Create a makeshift ordered list
    dummyList = [ x for x in range(1, 200, 3) ]
    # Print out the indices of all of the values in the list to compare with binary search 
    for idx, val in enumerate(dummyList):
        print(str(idx) + " : " + str(val))

    print() # Puts space between list and result

    # Perform binary search
    val = 166
    ans = binSearch(dummyList, val)
    # Print the index that the value was found at
    if ans > 0:
        print("The value %s was found at index: %s" %(val, ans))
    else:
        print("The value %s was not found in the list." %val)
    

def binSearch(arr, target):
    '''This method is used to demonstrate the use of the Binary Search algorithm
    in Python.'''

    start = 0            # First iteration means we are searching entire list
    end = len(arr) - 1   # Gets last index of array

    # Go through array until value is found or return -1
    while(start <= end):

        # Calculate midpoint
        mid =  int((start + end) / 2)

        if arr[mid] == target:  # Target found
            return mid
        elif target < arr[mid]: # Target is on the left half of the list
            end = mid - 1
        else:                   # Target is on the right half of the list
            start = mid + 1

    return (-1)   # Target is not in the list

    



if __name__ == "__main__":
    '''This file was not imported'''

    main()  # Start program